htdocs/product/stock/stockatdate.php

```diff
@@ -276,34 +276,109 @@
+if ($useStockSum) {
+	$sql .= ' SUM('.$stockField.') as '.$stockAlias.',';
+} else {
+	$sql .= ' '.$stockField.' as '.$stockAlias.',';
+}
+
+// --- Value fields (DRY implementation)
+// currentvalue and sellvalue are now built *once* using the $stockField variable.
+// This logic is no longer repeated three times.
+$sql .= ' SUM(COALESCE(ppe.pmp, p.pmp) * '.$stockField.') as currentvalue,';
+$sql .= ' SUM(p.price * '.$stockField.') as sellvalue';
+
+// Hooks select
 $parameters = array();
-$reshook = $hookmanager->executeHooks('printFieldListSelect', $parameters); // Note that $action and $object may have been modified by hook
+$reshook = $hookmanager->executeHooks('printFieldListSelect', $parameters);
 $sql .= $hookmanager->resPrint;
 
-$sql .= ' FROM '.MAIN_DB_PREFIX.'product as p';
+// --- FROM / JOIN
+// This is where the code block you provided in your last message begins.
+// Remember to apply the security (intval) and prefix ($db->prefix()) fixes here.
+$sql .= ' FROM '.$db->prefix().'product as p';
+
+// Join product_perentity si on peut en bénéficier (PMP par entité ou per-entity partagé)
+if (getDolGlobalString('MAIN_PRODUCT_PERENTITY_SHARED') || getDolGlobalString('MULTICOMPANY_PMP_PER_ENTITY_ENABLED')) {
+	$sql .= ' LEFT JOIN '.$db->prefix().'product_perentity as ppe'
+		.  ' ON ppe.fk_product = p.rowid AND ppe.entity = '.((int) $conf->entity);
+}
+
+// Stock joins
 if (!empty($search_fk_warehouse)) {
-	$sql .= ' LEFT JOIN '.MAIN_DB_PREFIX.'product_stock as ps ON p.rowid = ps.fk_product AND ps.fk_entrepot IN ('.$db->sanitize(implode(",", $search_fk_warehouse)).")";
+	// Case 1: warehouse filter -> ps
+	$warehouseList = array_map('intval', $search_fk_warehouse);
+	$sql .= ' LEFT JOIN '.$db->prefix().'product_stock as ps'
+		.  ' ON ps.fk_product = p.rowid'
+		.  ' AND ps.fk_entrepot IN ('.implode(',', $warehouseList).')';
+} elseif ($useSeparatedStock) {
+	// Case 2: sum via sp on visible warehouses (current entity + shared)
+	$entities = array_map('intval', explode(',', $visibleWarehousesEntities));
+	$sql .= ' LEFT JOIN '.$db->prefix().'product_stock as sp'
+		.  ' ON sp.fk_product = p.rowid'
+		.  ' AND sp.fk_entrepot IN (SELECT rowid FROM '.$db->prefix().'entrepot'
+		.  '                         WHERE entity IN ('.implode(',', $entities).'))';
 }
-// Add fields from hooks
+
+// Hooks join
 $parameters = array();
-$reshook = $hookmanager->executeHooks('printFieldListJoin', $parameters); // Note that $action and $object may have been modified by hook
+$reshook = $hookmanager->executeHooks('printFieldListJoin', $parameters);
 $sql .= $hookmanager->resPrint;
+
+// --- WHERE
 $sql .= ' WHERE p.entity IN ('.getEntity('product').')';
+
 if ($productid > 0) {
-	$sql .= " AND p.rowid = ".((int) $productid);
+	$sql .= ' AND p.rowid = '.((int) $productid);
 }
 if (!getDolGlobalString('STOCK_SUPPORTS_SERVICES')) {
-	$sql .= " AND p.fk_product_type = 0";
+	$sql .= ' AND p.fk_product_type = '.Product::TYPE_PRODUCT;
```

Comment:
SQL request must be forged using" for php string and ' inside for SQL string. Force cast of the constant Product::TYPE_PRODUCT into int to avoid false alertof the sqlinjection detector in CI.